pkg/ingestor/verifier/sigstore_verifier/sigstore_verifier_test.go

```diff
@@ -0,0 +1,393 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sigstore_verifier
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/key"
+	"github.com/guacsec/guac/pkg/ingestor/verifier"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/dsse"
+)
+
+type mockKeyProvider struct {
+	collector map[string]key.Key
+}
+
+func newMockProvider() *mockKeyProvider {
+	return &mockKeyProvider{
+		collector: map[string]key.Key{},
+	}
+}
+
+func (m *mockKeyProvider) RetrieveKey(id string) (*key.Key, error) {
+	if key, ok := m.collector[id]; ok {
+		return &key, nil
+	}
+	return nil, nil
+}
+
+func (m *mockKeyProvider) StoreKey(id string, pk *key.Key) error {
+	m.collector[id] = *pk
+	return nil
+}
+
+func (m *mockKeyProvider) DeleteKey(id string) error {
+	delete(m.collector, id)
+	return nil
+}
+
+func (m *mockKeyProvider) Type() key.KeyProviderType {
+	return "mock"
+}
+
+func setupOneProvider(t *testing.T) {
+	provider := newMockProvider()
+	key.RegisterKeyProvider(provider, "mock")
+
+	err := key.Store("SHA256:s9b/UAMASq9HN7RPBm5cIHQGoBOQA120kFdWLW/lT88", []byte(ecdsaPub), "mock")
+	if err != nil {
+		t.Fatal("failed to store into mock key provider")
+	}
+	err = key.Store("SHA256:843yiXZzbDfB0gA1snxYG5SISWMnDimw8/8Aew0nVNg", []byte(rsapub), "mock")
+	if err != nil {
+		t.Fatal("failed to store into mock key provider")
+	}
+}
+
+func randomData(t *testing.T, n int) []byte {
+	t.Helper()
+	rand.Seed(time.Now().UnixNano())
+	data := make([]byte, n)
+	if _, err := rand.Read(data[:]); err != nil {
+		t.Fatal(err)
+	}
+	return data
+}
+
+func TestSigstoreVerifier_Verify(t *testing.T) {
+	setupOneProvider(t)
+	foundECDSAKey, err := key.Find("SHA256:s9b/UAMASq9HN7RPBm5cIHQGoBOQA120kFdWLW/lT88")
```

Comment:
it looks like there's an implicit (non-encapsulating data format) contract between the key provider and the KeyID.... in this case, store should return a key ID so that it can be looked up.

Reply:
When storing the key, you pass in the keyID. In most cases the KeyID will be the hash of the key but it is not enforced.